### PR TITLE
feat(textfield) : add new component

### LINKED ISSUE #176: Add text field component
### Expected behavior 
Migrate https://fluentweb.com/components/text-field to FAST-DNA 

- Ensure style is maintained
- Ensure behavior is maintained
- Ensure accessibility is maintained

For testing follow https://github.com/Microsoft/fast-dna/wiki/the-process

### PR DESCRIPTION
**Pull request by emathiasen**
Closes #176

### CODE REVIEW

File: packages/fast-components-styles-msft/src/text-field/index.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { IDesignSystem } from "../design-system";
+import { ComponentStyles, ICSSRules } from "@microsoft/fast-jss-manager";
+import { ITextFieldClassNameContract } from "@microsoft/fast-components-class-name-contracts-base";
+import { toPx } from "@microsoft/fast-jss-utilities";
+import * as Chroma from "chroma-js";
+
+const styles: ComponentStyles<ITextFieldClassNameContract, IDesignSystem> = {
+    textField: {
+        color: "#757575",
```
Comment (nalogiudice) [inline packages/fast-components-styles-msft/src/text-field/index.ts]: You can use chroma to mix colors to get this result. See button for examples.

File: packages/fast-components-styles-msft/src/text-field/index.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        outline: `${toPx(1)} solid transparent`,
+        fontSize: toPx(14),
```
Comment (nalogiudice) [inline packages/fast-components-styles-msft/src/text-field/index.ts]: Should use applyType to add fontSize and lineHeight

File: packages/fast-components-styles-msft/src/text-field/index.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        fontWeight: "100",
+        "&:focus": {
+            outline: `${toPx(1)} solid #616161`,
+            border: `${toPx(1)} solid #616161`
+        },
+        "&:disabled": {
+            color: "#CCCCCC",
+            cursor: "not-allowed",
+            background: (config: IDesignSystem): string => {
+                return Chroma(config.backgroundColor).css();
```
Comment (nalogiudice) [inline packages/fast-components-styles-msft/src/text-field/index.ts]: You should just return `config.background` here - it is already in the correct format

File: packages/fast-components-styles-msft/src/text-field/index.ts (hunk 1 above)
Comment (nalogiudice) [reply packages/fast-components-styles-msft/src/text-field/index.ts]: This would go for all hardcoded hex values.

File: packages/fast-components-react-base/src/text-field/text-field.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+     * Generates class names
+     */
+    protected generateClassNames(): string {
+        return super.generateClassNames(get(this.props, "managedClasses.textField"));
+    }
+
+    /**
+     * Generates handled attributes based on props
+     */
+    private generateHandledAttributes(): Partial<ITextFieldHandledProps> {
```
Comment (nalogiudice) [inline packages/fast-components-react-base/src/text-field/text-field.tsx]: We are trying to get away from the `generateAttributes` method for a few reasons. The props can go directly on the render function.

File: packages/fast-components-react-base/src/text-field/text-field.props.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import * as React from "react";
+import { IManagedClasses, ITextFieldClassNameContract } from "@microsoft/fast-components-class-name-contracts-base";
+
+export enum TextFieldType {
+   date = "date",
```
Comment (nalogiudice) [inline packages/fast-components-react-base/src/text-field/text-field.props.ts]: @nicholasrice do we want to do it this way?

File: packages/fast-components-react-base/src/text-field/text-field.tsx (hunk 4 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/text-field/text-field.tsx]: See checkbox for examples

File: packages/fast-components-styles-msft/src/text-field/index.ts (hunk 1 above)
Comment (emathiasen) [reply packages/fast-components-styles-msft/src/text-field/index.ts]: Could this be moved into a utility? Seems like its going to be used frequently.

File: packages/fast-components-react-base/src/text-field/text-field.props.ts (hunk 5 above)
Comment (nicholasrice) [reply packages/fast-components-react-base/src/text-field/text-field.props.ts]: The enum looks right to me, is that what you're reffering to?

File: .github/ISSUE_TEMPLATE/Bug_report.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
+---
+name: Bug report
+about: Create a report to help us improve
+
```
Comment (nicholasrice) [inline .github/ISSUE_TEMPLATE/Bug_report.md]: I think you need to rebase to master - these files are already in. If you could do a round on this PR and ensure that only the intended files are changed that would be good, cause there are a bunch of seemingly unrelated changes to this PR.

File: packages/fast-components-styles-msft/src/text-field/index.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
+import { IDesignSystem } from "../design-system";
+import { ComponentStyles, ICSSRules } from "@microsoft/fast-jss-manager";
+import { ITextFieldClassNameContract } from "@microsoft/fast-components-class-name-contracts-base";
+import { toPx } from "@microsoft/fast-jss-utilities";
+import { applyType } from "../utilities/typography";
+import * as Chroma from "chroma-js";
+
+function applyMixedColor(incomingProperty: string, mixValue?: number, alpha?: number): ICSSRules<IDesignSystem> {
```
Comment (nicholasrice) [inline packages/fast-components-styles-msft/src/text-field/index.ts]: `incomingProperty` isn't really doing anything for this function but reducing extensibility. Instead, you could change to:

```
function applyMixedColor(minvalue?: number, alpha?: number): (config: IDesignSystem) => string {
    return function applyColors(config: IDesignSystem): string {
       if (alpha) {
           return Chroma.mix(config.foregroundColor, config.backgroundColor, mixValue).alpha(alpha).css();
       } else {
           return Chroma.mix(config.foregroundColor, config.backgroundColor, mixValue).css();
       }
   }
}
```

and later apply it:
```
{
    outlineColor: applyMixedColor(0.38)
}
```

Doing that would expand the scope of use of the function and reduce code complexity

File: .github/ISSUE_TEMPLATE/Bug_report.md (hunk 6 above)
Comment (emathiasen) [reply .github/ISSUE_TEMPLATE/Bug_report.md]: rebased with 68b3ee8

File: packages/fast-components-react-base/src/text-field/text-field.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * Renders the component
+     */
+    public render(): React.ReactElement<HTMLInputElement> {
+        return (
+            <input
+                {...this.unhandledProps()}
+                className={this.generateClassNames()}
+                disabled={this.props.disabled}
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/text-field/text-field.tsx]: `disabled={this.props.disabled || null}`

File: packages/fast-components-react-base/src/text-field/text-field.tsx [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+                placeholder={this.props.placeholder}
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/text-field/text-field.tsx]: `placeholder={this.props.placeholder || null}`

File: packages/fast-components-react-base/src/text-field/text-field.tsx [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    TextFieldType
+} from "./text-field.props";
+import { IManagedClasses, ITextFieldClassNameContract } from "@microsoft/fast-components-class-name-contracts-base";
+import { get } from "lodash-es";
+
+/* tslint:disable-next-line */
+class TextField extends Foundation<ITextFieldHandledProps & IManagedClasses<ITextFieldClassNameContract>, React.HTMLAttributes<HTMLInputElement>, {}> {
+    protected handledProps: HandledProps<ITextFieldHandledProps & IManagedClasses<ITextFieldClassNameContract>> = {
+        disabled: void 0,
+        error: void 0,
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/text-field/text-field.tsx]: If we've pulled error, can we pull it from here as well?

File: packages/fast-components-react-base/src/text-field/text-field.tsx [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        placeholder: void 0,
+        managedClasses: void 0,
+        type: void 0,
+        value: void 0
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/text-field/text-field.tsx]: value won't be passed if we are handling it but not applying it to the element.

If we don't plan on doing anything with it, let's remove this from here and props to ensure users can pass it as expected.

File: packages/fast-components-react-base/src/text-field/text-field.props.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+export interface ITextFieldHandledProps {
+    /**
+     * The disabled state
+     */
+    disabled?: boolean;
+
+    /**
+     * The error state
+     */
+    error?: boolean;
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/text-field/text-field.props.ts]: can we pull error from here as well

File: packages/fast-components-react-base/src/text-field/text-field.props.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * The text field input type
+     */
+    type?: TextFieldType;
+
+    /**
+     * The text field value
+     */
+    value?: string;
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/text-field/text-field.props.ts]: If we aren't handling this in the component, we can pull from props.

Comment (chrisdholt) [reply packages/fast-components-react-base/src/text-field/text-field.props.ts]: Don't forget to pull from schema's as well.

File: packages/fast-components-react-base/src/text-field/text-field.schema.json [hunk 14]
```diff
@@ -0,0 +1,37 @@
+{
+    "$schema": "http://json-schema.org/schema#",
+    "title": "Text Field",
```
Comment (nalogiudice) [inline packages/fast-components-react-base/src/text-field/text-field.schema.json]: Text field (de-capitalize field)

File: packages/fast-components-react-base/src/text-field/text-field.schema.json [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    "type": "object",
+    "id": "text-field",
+    "children": true,
+    "properties": {
+        "disabled": {
+            "title": "Disabled",
+            "type": "boolean"
+        },
+        "type": {
+            "title": "Text Field type",
```
Comment (nalogiudice) [inline packages/fast-components-react-base/src/text-field/text-field.schema.json]: Just "Type" - though this could be "Input type" as well

File: packages/fast-components-react-base/src/text-field/text-field.schema.json [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+                "search",
+                "tel",
+                "text",
+                "time",
+                "url",
+                "week"
+            ]
+        },
+        "placeholder": {
+            "title": "Text Field placeholder",
```
Comment (nalogiudice) [inline packages/fast-components-react-base/src/text-field/text-field.schema.json]: Just "Placeholder"

File: packages/fast-components-styles-msft/src/text-field/index.ts [hunk 17]
```diff
@@ -0,0 +1,45 @@
+import { IDesignSystem } from "../design-system";
+import { ComponentStyles, ICSSRules } from "@microsoft/fast-jss-manager";
+import { ITextFieldClassNameContract } from "@microsoft/fast-components-class-name-contracts-base";
+import { toPx } from "@microsoft/fast-jss-utilities";
+import { applyType } from "../utilities/typography";
+import * as Chroma from "chroma-js";
+
+function applyMixedColor(mixValue?: number, alpha?: number): (config: IDesignSystem) => string {
+    return function applyColors(config: IDesignSystem): string {
```
[the remark this answers is not in the document]
Comment (nalogiudice) [reply packages/fast-components-styles-msft/src/text-field/index.ts]: Not your concern, but we may want to make this a utility... thoughts @nicholasrice @chrisdholt ?

File: packages/fast-components-styles-msft/src/text-field/index.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+            return Chroma.mix(config.foregroundColor, config.backgroundColor, mixValue).css();
+        }
+    };
+}
+
+const styles: ComponentStyles<ITextFieldClassNameContract, IDesignSystem> = {
+    textField: {
+        ...applyType("t7", "vp1"),
+        color: applyMixedColor(0.46),
+        fontWeight: "100",
```
Comment (nalogiudice) [inline packages/fast-components-styles-msft/src/text-field/index.ts]: We have a utility for font weights that we should use

**Comment by nalogiudice** [comment]
Also, noticing there is no hover state

File: packages/fast-components-react-base/src/text-field/text-field.props.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
+import * as React from "react";
+import { IManagedClasses, ITextFieldClassNameContract } from "@microsoft/fast-components-class-name-contracts-base";
+
+export enum TextFieldType {
+   date = "date",
+   email = "email",
+   hidden = "hidden",
+   month = "month",
+   number = "number",
```
Comment (nalogiudice) [inline packages/fast-components-react-base/src/text-field/text-field.props.ts]: We usually have a separate component for number field, so I think this should be omitted... again thoughts?

File: packages/fast-components-styles-msft/src/text-field/index.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (awentzel) [reply packages/fast-components-styles-msft/src/text-field/index.ts]: Already filed issues from Code Climate to resolve this.  I should have notified you guys here.  This is a new feature just implemented last week. can be resolved at later date.

File: packages/fast-components-react-base/src/text-field/text-field.props.ts (hunk 19 above)
Comment (awentzel) [reply packages/fast-components-react-base/src/text-field/text-field.props.ts]: @nalogiudice : If you're saying that text-field should be one word, then should that also apply to the file name?  Also, I'm not understanding what you mean by Number?  Do you have an example or file we can look at?

File: packages/fast-components-styles-msft/src/text-field/index.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (chrisdholt) [reply packages/fast-components-styles-msft/src/text-field/index.ts]: Yep - we'll break this out in a separate task as we'll want to resolve other places this can be used.

File: packages/fast-components-react-base/src/text-field/text-field.props.ts (hunk 19 above)
Comment (chrisdholt) [reply packages/fast-components-react-base/src/text-field/text-field.props.ts]: With unhandled props I think we can apply all functionality we need to to ensure both text and number fields are supported with a single component. I know previously we did that, although I don't recall the exact reason. Is there a specific scenario you can think of that would require two different components here?

**Comment by emathiasen** [comment]
@nalogiudice [Figma](https://www.figma.com/file/Kc8GyAEJw6EGBQFlEvrsnsq7/FAST-Components---MS) comp didn't show a hover state.. We can definitely add one if other components have a similar interaction. 
Looping in @Falkicon should add a hover state for the Text field component?

File: packages/fast-components-styles-msft/src/text-field/index.ts (hunk 18 above)
Comment (emathiasen) [reply packages/fast-components-styles-msft/src/text-field/index.ts]: JSS expecting fontWeight as a string. Should utility be updated to reflect this or should it be converted in the components handling of that property?

File: packages/fast-components-react-base/src/text-field/text-field.props.ts (hunk 19 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/text-field/text-field.props.ts]: I think we may be able to get by with unhandled props. Perhaps an example that tests that would be beneficial in this case. I.E. uses min, max, step etc... to verify it works as expected.

@awentzel if you look on fluentweb you'll see we have a number field component and a text field component, as separate components. But, @chrisdholt 's point is valid I believe, with my caveat above.

Text field is two words, no hyphen

File: packages/fast-components-styles-msft/src/text-field/index.ts (hunk 18 above)
Comment (nalogiudice) [reply packages/fast-components-styles-msft/src/text-field/index.ts]: Interesting. Not really in the scope of the PR but we could either simply `toString` the functions' results, or change it to return a string. I wonder if it is being used elsewhere and needs to be a number in that case.

Comment (nicholasrice) [reply packages/fast-components-styles-msft/src/text-field/index.ts]: we may just need to update the typings for a stylesheet - I think right now it explicitly expects a string but sometimes numbers are valid.

File: packages/fast-components-react-base/src/text-field/text-field.tsx [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+    ITextFieldUnhandledProps,
+    TextFieldType
+} from "./text-field.props";
+import { IManagedClasses, ITextFieldClassNameContract } from "@microsoft/fast-components-class-name-contracts-base";
+import { get } from "lodash-es";
+
+/* tslint:disable-next-line */
+class TextField extends Foundation<ITextFieldHandledProps & IManagedClasses<ITextFieldClassNameContract>, React.HTMLAttributes<HTMLInputElement>, {}> {
+    protected handledProps: HandledProps<ITextFieldHandledProps & IManagedClasses<ITextFieldClassNameContract>> = {
+        disabled: void 0,
```
[the remark this answers is not in the document]
Comment (nalogiudice) [reply packages/fast-components-react-base/src/text-field/text-field.tsx]: @chrisdholt has this been resolved?

File: packages/fast-components-react-msft/src/text-field/examples.data.ts [hunk 21]
```diff
@@ -0,0 +1,38 @@
+import { ISnapshotTestSuite } from "@microsoft/fast-jest-snapshots-react";
+import { IManagedClasses } from "@microsoft/fast-jss-manager-react";
+import TextField from "./index";
+import { ITextFieldHandledProps, TextFieldType } from "@microsoft/fast-components-react-base";
+import * as schema from "@microsoft/fast-components-react-base/dist/text-field/text-field.schema.json";
+
+export default {
+    name: "text-field",
```
Comment (nalogiudice) [inline packages/fast-components-react-msft/src/text-field/examples.data.ts]: Should be `text field` as well
